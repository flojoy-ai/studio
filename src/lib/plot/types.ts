export interface Drawable {
  draw: () => void;
}
