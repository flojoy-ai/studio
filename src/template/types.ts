export interface SomeObj {
    id: string,
    propertyX: string;
    propertyY: string;
}