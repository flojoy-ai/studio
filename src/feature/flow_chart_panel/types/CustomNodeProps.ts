export type ElementsData = {
  id: string;
  label: string;
  func: string;
  type: string;
  running?: boolean;
  failed?: boolean;
  ctrls: {
    [key: string]: {
      functionName: string;
      param: string;
      value: string | boolean | number;
    };
  };
  inputs?: Array<{ name: string; id: string; type: string }>;
  selects?: any;
  selected?: boolean;
  handleRemove?: (nodeId: string) => void;
};

export interface CustomNodeProps {
  data: ElementsData;
}
