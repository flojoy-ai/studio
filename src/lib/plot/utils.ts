export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
