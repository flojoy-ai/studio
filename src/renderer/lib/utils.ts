import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
