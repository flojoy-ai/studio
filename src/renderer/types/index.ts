export type {
  ALL_DC_TYPE,
  DataContainer,
  DataContainerType,
} from "./dataContainer";
export type { CustomNodeProps, BlockData } from "./node";
export type {
  DataContainer2PlotlyProps,
  OverridePlotData,
  PlotProps,
} from "./plotly";
