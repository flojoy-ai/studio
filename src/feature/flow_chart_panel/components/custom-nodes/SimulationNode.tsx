import HandleComponent from "@feature/flow_chart_panel/components/HandleComponent";
import { CustomNodeProps } from "@feature/flow_chart_panel/types/CustomNodeProps";
import { useFlowChartState } from "@hooks/useFlowChartState";
import { Text, Box, clsx, createStyles } from "@mantine/core";
import { memo } from "react";
import { useNodeStyles } from "../DefaultNode";
import NodeWrapper from "../NodeWrapper";

const useStyles = createStyles((theme) => {
  const accent =
    theme.colorScheme === "light"
      ? theme.colors.accent1[0]
      : theme.colors.accent2[0];
  return {
    simulationNode: {
      width: 130,
      borderRadius: 6,
      flexDirection: "column",
      justifyContent: "center",
      border: `1px solid ${accent}`,
      color: accent,
      backgroundColor: accent + "27",
      textAlign: "center",
      overflowWrap: "anywhere",
      padding: "16px",
    },
  };
});

const SimulationNode = ({ data, handleRemove }: CustomNodeProps) => {
  const nodeClasses = useNodeStyles().classes;
  const { classes } = useStyles();
  const { runningNode, failedNode } = useFlowChartState();
  const params = data.inputs ?? [];

  let selectShadow = "";
  if (runningNode === data.id || data.selected) {
    selectShadow =
      data.func === "LINSPACE"
        ? nodeClasses.defaultShadow
        : nodeClasses.simulationShadow;
  }

  return (
    <NodeWrapper data={data} handleRemove={handleRemove}>
      <Box
        className={clsx(
          selectShadow,
          failedNode === data.id ? nodeClasses.failShadow : ""
        )}
      >
        <Box
          className={clsx(nodeClasses.nodeContainer, classes.simulationNode)}
          sx={{
            ...(params.length > 0 && { padding: "0px 0px 8px 0px" }),
          }}
        >
          <Box data-testid="data-label-design">
            <Text
              weight={600}
              size="xl"
              sx={{ letterSpacing: 1, fontFamily: "Open Sans" }}
            >
              {data.label}
            </Text>
          </Box>
          <Box
            display="flex"
            h={params.length > 0 ? (params.length + 1) * 40 : "fit-content"}
            sx={{
              flexDirection: "column",
            }}
          >
            <HandleComponent data={data} />
          </Box>
        </Box>
      </Box>
    </NodeWrapper>
  );
};

export default memo(SimulationNode);
