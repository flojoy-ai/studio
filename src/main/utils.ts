import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
