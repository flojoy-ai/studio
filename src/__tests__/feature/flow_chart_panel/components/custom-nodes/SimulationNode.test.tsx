import SimulationNode from "@src/feature/flow_chart_panel/components/custom-nodes/SimulationNode";
import { CustomNodeProps } from "@feature/flow_chart_panel/types/CustomNodeProps";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

const props: CustomNodeProps = {
  data: {
    id: "test-id",
    label: "test",
    func: "test",
    type: "test",
    ctrls: {},
    inputs: [
      {
        id: "test",
        name: "test",
        type: "test",
      },
    ],
  },
};

jest.mock("@src/hooks/useFlowChartState");

jest.mock("@feature/flow_chart_panel/components/HandleComponent", () => {
  const mockChildren = jest
    .fn()
    .mockReturnValue(<div data-testid="handle-component" />);
  return { __esModule: true, default: mockChildren };
});

jest.mock("@src/hooks/useSocket");

describe("ConditionalNode", () => {
  it("checks the snapshot", () => {
    const { container } = renderWithTheme(<SimulationNode {...props} />);
    expect(container).toMatchSnapshot();
  });
  it("should contain the data label text", () => {
    const { container, getByText } = renderWithTheme(
      <SimulationNode {...props} />
    );
    const textField = getByText(props.data.label);
    expect(textField).toBeInTheDocument();
  });
  it("check if the handle component rendered", () => {
    const { getByTestId } = renderWithTheme(<SimulationNode {...props} />);
    const component = getByTestId("handle-component");
    expect(component).toBeInTheDocument();
  });
});
