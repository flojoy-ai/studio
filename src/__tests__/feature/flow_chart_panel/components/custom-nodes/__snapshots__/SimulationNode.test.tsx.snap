// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionalNode checks the snapshot 1`] = `
<div>
  <div
    class="mantine-n8y7xk"
    data-testid="node-wrapper"
  >
    <div
      class="mantine-1avyp1d"
    >
      <div
        class="mantine-1ee35f4"
      >
        <div
          class="mantine-1avyp1d"
          data-testid="data-label-design"
        >
          <div
            class="mantine-1avyp1d"
          >
            test
          </div>
        </div>
        <div
          class="mantine-1b6b4my"
        >
          <div
            data-testid="handle-component"
          />
        </div>
      </div>
    </div>
  </div>
  ,
</div>
`;
